src/main/java/io/github/ImpactDevelopment/installer/optifine/OptiFine.java

```diff
@@ -0,0 +1,213 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package io.github.ImpactDevelopment.installer.optifine;
+
+import javax.annotation.Nullable;
+import java.io.*;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Enumeration;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+
+public class OptiFine {
+
+    private static final String GROUP = "optifine";
+    private static final Pattern TWEAKER_REGEX = Pattern.compile("^TweakClass:\\s+(.+)$");
+
+    private final Path jarPath;
+    private final String[] version;
+    private final String tweaker;
+    private final String transformer;
+    private final String launchwrapperVersion;
+
+    // patcherMethod takes a vanilla "base" jar along with an OptiFine "diff" jar and writes the result to a "mod" jar
+    // public static void process(File baseFile, File diffFile, File modFile)
+    private final Method patcherMethod;
+
+    public OptiFine(Path jarPath) throws RuntimeException {
+        // Set locals first while iterating the zip file, then set the final fields after
+        String version;
+        String launchwrapperVersion = null;
+        String tweaker = "";
+        String transformer = "";
+
+        // load required methods from the OptiFine installer jar
+        Method getOptiFineVersionMethod;
+        Method patcherMethod;
+        try {
+            ClassLoader classloader = new URLClassLoader(new URL[]{jarPath.toUri().toURL()});
+            patcherMethod = classloader.loadClass("optifine.Patcher").getMethod("process", File.class, File.class, File.class);
+            getOptiFineVersionMethod = classloader.loadClass("optifine.Installer").getMethod("getOptiFineVersion", ZipFile.class);
+        } catch (MalformedURLException | ClassNotFoundException | NoSuchMethodException e) {
+            throw new RuntimeException("Unable to load OptiFine classes", e);
+        }
+
+        // Iterate over the zip entries in the jar and extract any info we care about
+        try {
+            try (ZipFile jar = new ZipFile(jarPath.toFile())) {
+                // Extract optifine version
+                version = (String) getOptiFineVersionMethod.invoke(null, jar);
+
+                final Enumeration<? extends ZipEntry> entries = jar.entries();
+                while (entries.hasMoreElements()) {
```

Comment:
Can't you foreach this?

Reply:
Don't think so. `Enumeration` only provides `hasMoreElements` and `nextElement` methods, no `forEach` or `stream`. A lambda would also make it harder to set the locals variables wouldn't it?

Reply:
[This](http://java-performance.info/how-to-iterate-zip-file-records/) is why I used `ZipFile` over `java.util.zip.`ZipInputStream`.